### PR TITLE
ARROW-6065: [C++][Parquet] Clean up parquet/arrow/reader.cc, reduce code duplication, improve readability

### PR DESCRIPTION
**Pull request by wesm**
This is strictly a refactoring PR. I'm going to start working (for a new PR) on some refactoring of the handling of schemas and nested types (which is also pretty messy in my opinion). The motivation for this is to be able to more cleanly reason about direct dictionary-decoding without having to resort to such hacks as the current `FixSchemas` function

Also cleans up Parquet includes using IWYU

### CODE REVIEW

**Comment by wesm** [comment]
@emkornfield I'm also trying to clean up some business logic in such a way that it will help you with implementing nested reads, though there is little enough logic related to nested data that you might want to start fresh anyway

File: cpp/src/parquet/arrow/arrow-reader-writer-test.cc [hunk 1]
```diff
@@ -1845,8 +1845,8 @@ TEST(TestArrowReadWrite, MultithreadedRead) {
 }
 
 TEST(TestArrowReadWrite, ReadSingleRowGroup) {
-  const int num_columns = 20;
-  const int num_rows = 1000;
+  const int num_columns = 10;
+  const int num_rows = 100;
```
Comment (pitrou) [inline cpp/src/parquet/arrow/arrow-reader-writer-test.cc]: Is this to make the test faster?

File: cpp/src/parquet/arrow/reader.cc [hunk 2]
```diff
@@ -216,1462 +184,488 @@ class RowGroupRecordBatchReader : public ::arrow::RecordBatchReader {
 };
 
 // ----------------------------------------------------------------------
-// File reader implementation
+// FileReaderImpl forward declaration
```
Comment (pitrou) [inline cpp/src/parquet/arrow/reader.cc]: It's not really a "forward declaration", right? Seems like an implementation to me :-)

File: cpp/src/parquet/arrow/reader.cc [hunk 3]
```diff
@@ -216,1462 +184,488 @@ class RowGroupRecordBatchReader : public ::arrow::RecordBatchReader {
...
+  }
+
+  FileColumnIteratorFactory AllRowGroupsFactory() {
+    return SomeRowGroupsFactory(AllRowGroups());
+  }
+
+  int64_t GetTotalRecords(const std::vector<int>& row_groups, int column_chunk = 0) {
+    // Can throw exception
+    int64_t records = 0;
+    for (int j = 0; j < static_cast<int>(row_groups.size()); j++) {
```
Comment (pitrou) [inline cpp/src/parquet/arrow/reader.cc]: Use a range-based loop? `for (const auto& row_group : row_groups) ...`

File: cpp/src/parquet/arrow/reader.cc [hunk 4]
```diff
@@ -1790,40 +784,195 @@ Status StructImpl::NextBatch(int64_t records_to_read,
...
+    std::vector<std::future<Status>> futures;
+    auto pool = ::arrow::internal::GetCpuThreadPool();
+    for (int i = 0; i < num_fields; i++) {
+      futures.push_back(pool->Submit(ReadColumnFunc, i));
+    }
+    Status final_status = Status::OK();
+    for (auto& fut : futures) {
+      Status st = fut.get();
+      if (!st.ok()) {
+        final_status = std::move(st);
```
Comment (pitrou) [inline cpp/src/parquet/arrow/reader.cc]: Or simply `final_status &= fut.get()`.

File: cpp/src/parquet/arrow/reader.h [hunk 5]
```diff
@@ -201,117 +192,78 @@ class PARQUET_EXPORT FileReader {
...
- private:
-  FileReader(::arrow::MemoryPool* pool, std::unique_ptr<ParquetFileReader> reader,
-             const ArrowReaderProperties& properties);
-
-  friend ColumnChunkReader;
-  friend RowGroupReader;
-
-  class PARQUET_NO_EXPORT Impl;
-  std::unique_ptr<Impl> impl_;
+  virtual ~FileReader() = default;
```
Comment (pitrou) [inline cpp/src/parquet/arrow/reader.h]: Hmm... I'm not sure it's ok to use `= default` on a virtual destructor of a DLL-exported class. I think it's safer to define an empty destructor explicitly in the `.cc` file, though I may be mistaken.

File: cpp/src/parquet/arrow/reader.h [hunk 6]
```diff
@@ -201,117 +192,78 @@ class PARQUET_EXPORT FileReader {
...
 };
 
-class PARQUET_EXPORT RowGroupReader {
+class RowGroupReader {
```
Comment (pitrou) [inline cpp/src/parquet/arrow/reader.h]: Do we need to PARQUET_EXPORT this class and also `ColumnChunkReader`?

File: cpp/src/parquet/arrow/reader_internal.cc [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+  RETURN_NOT_OK(::arrow::AllocateBuffer(pool, buffer_size, &data));
+
+  // Transfer boolean values to packed bitmap
+  auto values = reinterpret_cast<const bool*>(reader->values());
+  uint8_t* data_ptr = data->mutable_data();
+  memset(data_ptr, 0, buffer_size);
+
+  for (int64_t i = 0; i < length; i++) {
+    if (values[i]) {
+      ::arrow::BitUtil::SetBit(data_ptr, i);
```
Comment (pitrou) [inline cpp/src/parquet/arrow/reader_internal.cc]: Should this use `GenerateBitsUnrolled` for higher perf?

File: cpp/src/parquet/arrow/reader_internal.cc [hunk 8]
```diff
@@ -0,0 +1,616 @@
...
+  return Status::OK();
+}
+
+Status TransferBinary(RecordReader* reader,
+                      const std::shared_ptr<DataType>& logical_value_type,
+                      std::shared_ptr<ChunkedArray>* out) {
+  if (reader->read_dictionary()) {
+    return TransferDictionary(reader, logical_value_type, out);
+  }
+  auto binary_reader = dynamic_cast<internal::BinaryRecordReader*>(reader);
```
Comment (pitrou) [inline cpp/src/parquet/arrow/reader_internal.cc]: Use `checked_cast`?

File: cpp/src/parquet/arrow/reader_internal.cc [hunk 9]
```diff
@@ -0,0 +1,616 @@
...
+  for (int i = 0; i < data.num_chunks(); ++i) {
+    string_chunks.push_back(ShallowCast(*data.chunk(i), logical_value_type));
+  }
+  return std::make_shared<ChunkedArray>(string_chunks);
+}
+
+Status TransferDictionary(RecordReader* reader,
+                          const std::shared_ptr<DataType>& logical_value_type,
+                          std::shared_ptr<ChunkedArray>* out) {
+  auto dict_reader = dynamic_cast<internal::DictionaryRecordReader*>(reader);
```
Comment (pitrou) [inline cpp/src/parquet/arrow/reader_internal.cc]: Use `checked_cast`?

**Review by pitrou — COMMENTED** [review]
I didn't really look at implementation details. A couple comments below.

File: cpp/src/parquet/arrow/arrow-reader-writer-test.cc (hunk 1 above)
Comment (wesm) [reply cpp/src/parquet/arrow/arrow-reader-writer-test.cc]: Yes, and easier to debug

File: cpp/src/parquet/arrow/reader.cc (hunk 4 above)
Comment (wesm) [reply cpp/src/parquet/arrow/reader.cc]: This is copy-paste

File: cpp/src/parquet/arrow/reader.h (hunk 6 above)
Comment (wesm) [reply cpp/src/parquet/arrow/reader.h]: Nope

File: cpp/src/parquet/arrow/reader.h (hunk 5 above)
Comment (wesm) [reply cpp/src/parquet/arrow/reader.h]: Hm. Would be nice to know what the C++ standard says

File: cpp/src/parquet/arrow/reader_internal.cc (hunk 7 above)
Comment (wesm) [reply cpp/src/parquet/arrow/reader_internal.cc]: Old code

File: cpp/src/parquet/arrow/reader_internal.cc (hunk 8 above)
Comment (wesm) [reply cpp/src/parquet/arrow/reader_internal.cc]: `static_cast` doesn't work with this type

File: cpp/src/parquet/arrow/reader_internal.cc (hunk 9 above)
Comment (wesm) [reply cpp/src/parquet/arrow/reader_internal.cc]: `static_cast` doesn't work with this type

**Comment by wesm** [comment]
I'll address a couple comments in my follow up patch. I'm going to merge this so I'm not stacking up patches

File: cpp/build-support/iwyu/mappings/arrow-misc.imp [hunk 10]
```diff
@@ -27,6 +27,7 @@
   { include: ["<ext/alloc_traits.h>", private, "<unordered_map>", public ] },
   { include: ["<ext/alloc_traits.h>", private, "<unordered_set>", public ] },
   { include: ["<ext/alloc_traits.h>", private, "<vector>", public ] },
+  { include: ["<bits/exception.h>", private, "<exception>", public ] },
```
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/build-support/iwyu/mappings/arrow-misc.imp]: yeah, see another project e.g. https://github.com/apache/kudu/tree/master/build-support/iwyu/mappings

the trouble is that IWYU in some cases will find the "minimal" header to obtain certain symbols which might be something internal to the STL implementation, but that will vary on different compilers (e.g. MSVC's internal STL stuff may be different) so you need to force it in some cases to use the right "official"/"public" headers

**Comment by emkornfield** [comment]
@wesm thanks, I've been a little delayed with the parquet stuff and this week, I'd like to try to knock off java/c++ compatibility since it seems you still have a few more patches to do.

**Comment by wesm** [comment]
@emkornfield yes, I'll let you know when it's the "all clear" viz-a-viz refactoring

**Pull request closed without merging**